Avoid reusing subnets when allocating from pools

ctelfer opened the pull request:
This patch changes the algorithm for selecting the next pool to allocate from an address space.   Previously, the list of potential subnets was fixed and the allocator always searched for an available subnet starting from the beginning of the list.  This means that as soon as libnetwork frees a network, it will likely use the same subnet range for the next network that it allocates.   There is nothing wrong with this in principle, but it does have the effect of making debugging a bit more tedious at times when subnets overlap frequently in the logs.

This patch keeps a cursor per address space pointing to the subnet to consider next for allocation and on each successful search of the predefined pools, sets it to the the following subnet.  The algorithm is robust against changing predefined pool lists, resetting the cursor to zero whenever it is found to be outside of the valid range of pool indices.  The patch does not attempt to reset roll back the cursor should the allocator select a subnet, but then fail to allocate it for other reasons.  This means that it is possible for some subnets to be *temporarily* skipped in the face of other allocation errors.   However the cursor will eventually wrap back around and these subnets will get a chance for allocation at that point.

This patch also includes a unit test to confirm the new behavior.

selansen commented on ipam/allocator_test.go:
```diff
@@ -590,6 +590,45 @@ func TestGetSameAddress(t *testing.T) {
+
+		// First get all pools until they are exhausted to
+		pList := []string{}
+		pool, _, _, err := a.RequestPool(localAddressSpace, "", "", nil, false)
```
I feel RequestPool call can be simplified like below
pool, _, _, err := a.RequestPool(localAddressSpace, "", "", nil, false)====> remove this line 
		for err == nil {====> this will be true first time anyways.
			pool, _, _, err = a.RequestPool(localAddressSpace, "", "", nil, false)
                        pList = append(pList, pool)
		}

ctelfer replied:
Ah, but I don't want to append the value pool (which may be an empty string) to the list when err != nil. That would be wrong. So I'd need a test for err inside the body of the loop. But that's 2 extra lines to mirror the loop test. :(

fcrisciani commented on ipam/allocator.go:
```diff
@@ -397,20 +402,34 @@ func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (*net.IPNet, error)
+	aSpace.Lock()
+	defer aSpace.Unlock()
+
+	c := a.asIndices[as]
```
maybe silly but can we give a more meaningful name to this variable?
something like addressSpaceIndex or nextNetworkIdx?

ctelfer replied:
Are you referring to `c` or to `asIndices`?  If the former, I'd chafe a bit at such a long name for an index variable.  If the latter, I'd have expected the comment to be at its declaration.  Hence the confusion.

fcrisciani replied:
I was referring to `c`

fcrisciani commented on ipam/allocator.go:
```diff
@@ -397,20 +402,34 @@ func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (*net.IPNet, error)
+		c = 0
+	}
+
+	for i := 0; i < len(pdl); i, c = i+1, c+1 {
```
actually I would use the `%`.

if you remove line 430 and at line 413 you put: `c = (c+1)% len(pdl)` you can remove also the if 412-415 (not sure if you have to handle the case of len(pdl) == 0 or is already handled before so cannot happen.

at the end of the loop or on the top then you can still use the same modulo format for the increment

ctelfer replied:
Go doesn't have a comma operator, so you can't do `i++,c++`. 

Re: removing line 430 -- that's the line that saves the cursor so that it's state is kept around for the next call.  Can't remove it. 

Yes, could use modulo, but I was coding defensively against the possibility that the list could change size.  As such there had to be *something* that kept `c <= len(pdf)` in the *first* iteration.  Yes, it could have been written as:
```
for i, c := 0, c%len(pdl); i < len(pdl); i, c = i+1, (c+1)%len(pdl) {
...
```
But that looks even messier.  I was striving for clarity.  (and if performance were a concern: these days a small branch usually costs much less than a modulo operation)

fcrisciani replied:
sorry I meant to save the value of c and not the next value, so that when you start the for loop you can init it with (c+1)% new length.
Also I was thinking that because there is a lock the length should remain fixed right?

ctelfer replied:
Ah, ok.  I understand now.

Re: the lock ... I was really referring to a change in the length of the pools between calls to getPredefinedPool() ...  In practice that currently never happens because the list of pools gets set once at initialization time and then left alone.  So, again, much more defensive than it probably needs to be.

Looking at the code, I realize that I'd assumed that getPredefineds() returned a copy of the list for some good reason like being called in multiple places or being called in the face of changes.  But it only ever gets invoked in getPredefinedPool.  So maybe it is worth getting rid of entirely.

Actually, above there was a specific reason that I *didn't* do `c := (a.asIndices[as]+1) % len(pdl)` or `for i, c := 0, (a.asIndices[as]+1)%len(pdl) ; ... {`   It was even more defensiveness.   If the length of the predefined pools were 0 for any reason (i.e. the list was empty) then this would through an integer exception.

ctelfer replied:
Given the above, it's apparent that the code could be clearer.   Would the following be ok?  (comments not to go in real code)

```
// REMOVE getPredefineds() entirely
...
aSpace.Lock()
defer aSpace.Unlock()

predefined := a.predefined[as]   // avoid map lookup per reference
for i, pi := 0, a.predefinedStartIdx[as]; i < len(predefined); i, pi := i+1,pi+1 {
    if pi >= len(predefined) {    // no test for pi < 0, but can only happen with corruption and we'll get a noisy panic if it occurs
        pi = 0
    } 
    nw := predefined[pi]
    ...
        a.predefinedStartIdx[as] = pi + 1
        return nw, nil
}
```

ctelfer commented on ipam/allocator.go:
```diff
@@ -397,20 +402,34 @@ func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (*net.IPNet, error)
+	c := a.asIndices[as]
+
+	// Defensive and redundant unless there is a bug
+	if c < 0 {
```
I'm not sure it should be *that* defensive.  I've half a mind to take it out entirely, TBH.  We could really get ourselves in a twist trying to code for every possible condition like that and adding warning messages for all of them.

selansen commented:
LGTM

fcrisciani commented:
Talked with @ctelfer offline he is following up with another patch to see if we can clean up further, will wait for that

ctelfer commented:
Updated the patch.  At first I was going to remove the getPredefineds() call and make the predefined networks a property of the addrSpace structure since that seems logical.  However, it became apparent why the predefined networks were kept separate:  addrSpaces is really designed to be a serialized object tracking the state of each address space.  Serializing the predefined pools and search index would not only be cumbersome, but actually could be considered wrong since it would restore previous predefined networks when they could be overridden with a new configuration.

So instead this patch embraces the getPredefined() and returns a copy of the predefined list *except* that it rotates the copy so that it starts with the start index element.  I also added a helper function to then update the start element index.   This keeps the locking clean by preserving the property that the code never holds the Allocator and addrSpace locks simultaneously.  It also makes the search loop simpler.

The unit test remains unchanged and still passes.

ctelfer commented:
Argh.  Other tests failing, however.  Checking why.

ctelfer commented:
Found the issue, fixed and re-pushed.

fcrisciani commented on ipam/allocator.go:
```diff
@@ -397,21 +420,23 @@ func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (*net.IPNet, error)
 			continue
 		}
-		aSpace.Lock()
 		if _, ok := aSpace.subnets[SubnetKey{AddressSpace: as, Subnet: nw.String()}]; ok {
```
can we add a comment that this check verifies that the subnet is already in use

ctelfer replied:
Sure, will do.

fcrisciani commented on ipam/allocator.go:
```diff
@@ -397,21 +420,23 @@ func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (*net.IPNet, error)
-			aSpace.Unlock()
 			continue
 		}
 		if !aSpace.contains(as, nw) {
```
This check not sure what is trying to achieve, I would expect that the network coming from the predefined pool is always non overlapping. Any idea of the real purpose for this?

ctelfer replied:
hrm....  Not sure...   Maybe there was some concern when doing the user-defined pools that one might have declared overlapping pools?

abhi commented on ipam/allocator_test.go:
```diff
@@ -590,6 +590,45 @@ func TestGetSameAddress(t *testing.T) {
+		// Now try to allocate then free nPool pools sequentially.
+		// Verify that we don't see any repeat networks even though
+		// we have freed them.
+		seen := map[string]bool{}
```
I think this will always succeed no ?

abhi replied:
ignore. It makes sense.

ctelfer commented on ipam/allocator.go:
```diff
@@ -374,13 +380,30 @@ func (a *Allocator) retrieveBitmask(k SubnetKey, n *net.IPNet) (*bitseq.Handle,
-	for _, pool := range a.predefined[as] {
-		l = append(l, pool)
+
+	p := a.predefined[as]
```
I like the use of append.  Will incorporate.   But don't want to use `%` because the list could, in theory, be empty leading to a divide-by-zero exception.

abhi replied:
I think that safety check can be added as if before these 3 lines. It just keeps the code simple and short.

ctelfer commented on ipam/allocator.go:
```diff
@@ -374,13 +380,30 @@ func (a *Allocator) retrieveBitmask(k SubnetKey, n *net.IPNet) (*bitseq.Handle,
 
+func (a *Allocator) updateStartIndex(as string, amt int) {
+	a.Lock()
+	i := a.predefinedStartIndices[as] + amt
```
It should never be < 0.  It's purely defensive.   So is the avoidance of modulo: see above.

ctelfer commented on ipam/allocator.go:
```diff
@@ -397,21 +420,23 @@ func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (*net.IPNet, error)
-	for _, nw := range a.getPredefineds(as) {
+	predefined := a.getPredefineds(as)
+
+	aSpace.Lock()
```
Actually, this is kind of what the original locking was before I touched it.  I did a deferred unlock originally when I did this patch.  However, on review of the code I found that the code avoids ever holding both an addrSpace lock and an allocator lock at the same time.   Making the addrSpace lock deferred will hold it through the updateStartIndex() which will perform an allocator lock.   It *probably* won't lead to a deadlock?  But maintaining the discipline of never holding both types of locks simultaneously makes it a guarantee.

dmaring commented:
This feature will be critical to Visa as they scale their services on a single overlay network.  Visa recently notified us of their plans to scale their production environment (Horizon) from 300 containers on 36 nodes /23 network to 600+ containers on 60 nodes on /22 network.  Based on testing by the cambridge team, its suggested that this will help manage risk with their scale.

https://docker.slack.com/archives/CA6M1TY7Q/p1534329477000100
